core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -402,8 +412,8 @@ class GroupCoordinator(val brokerId: Int,
             responseCallback(memberMetadata.assignment, Errors.NONE)
             completeAndScheduleNextHeartbeatExpiration(group, group.get(memberId))
 
-          case Dead =>
-            throw new IllegalStateException(s"Reached unexpected condition for Dead group ${group.groupId}")
+          case _ =>
```

Comment:
Some cleanup here for code generality.

Reply:
Nice one

Reply:
I don't think compiler would complain even if we do not exhaust all the cases in `match`. On the other hand, my rationale is that we use `_` for all cases which we would not expect (i.e. it would be a bug) and then throw. This is safer to any possible new state that we add without the handling logic.

Reply:
The compiler gives a warning if the match is not exhaustive. Hopefully some day we can make compiler warnings fatal. In any case, if the match is not exhaustive, you would still get an error if there is no match, so is there a difference?

Reply:
Okay, I think that means we can just remove all the `case _ => throws` as it would throw still then?

Reply:
Yeah, what @hachikuji said is correct. We should only use a catch all if the logic is not meant to be exhaustive. Here, it makes sense to handle the Dead case instead of catch all and we will get compiler warnings if additional cases are added.